### ISSUE TITLE
Eclipse Oxygen very slow server startup when debug on GAE projects

### ISSUE THREAD

**Issue by aleguerrini** [GoogleCloudPlatform/google-cloud-eclipse#2727]
I have a Java GAE (Google App Engine) project in Eclipse Oxygen Release 4.7.2 and Google Cloud Tools for Eclipse plugin version 1.5.0.

I have simple a Servlet class in my code:

```
package it.ale.test;

import java.io.IOException;

import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

@WebServlet(name = "HelloAppEngine", urlPatterns = { "/hello" }, loadOnStartup = 1)
public class HelloAppEngine extends HttpServlet {

    @Override
    public void init() throws ServletException {

        super.init();

        System.out.println("Print something.");

    }

    @Override
    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {

        response.setContentType("text/plain");
        response.setCharacterEncoding("UTF-8");

        response.getWriter().print("Hello App Engine!\r\n");

    }

}
```

when I put a **debug breakpoint** on any line of the _"public void init()"_ function, example on the line:

`System.out.println("Print something.");`

and I start the server in DEBUG mode **(Debug As -> App Engine)**, the server startup is incredible SLOW and I reach the error message:

> server app engine Standard at localhost was unable to start within 240 seconds. If the server requires more time, try increasing the timeout in the server editor.

**Comment by briandealwis** [comment]
I just responded to your post on StackOverflow.  Does this only happen when running with _Debug_?  Do you see it if you use _Run As_?  Does disabling breakpoints (_Run > Skip All Breakpoints_) help?

**Comment by aleguerrini** [comment]
Thanks for your response. 
I answer you on the StackOverflow ... alternatively if you prefer I continue the conversation here.

**Comment by briandealwis** [comment]
I didn't realize your Eclipse freezes, and I can reproduce this too — my entire Eclipse freezes.

**Comment by aleguerrini** [comment]
Thanks Brian. I hope the information I have provided you may be sufficient. If you need more information or logs let me know.

**Comment by briandealwis** [comment]
@aleguerrini do you have an exception breakpoint on `NullPointerException`?  I do, and when I disable that exception breakpoint then all is well.

**Comment by aleguerrini** [comment]
No. The problem occurs when I put the breakpoint on this line:

`System.out.println("Print something.");`

on init function of the servlet of the example code above.

**Comment by briandealwis** [comment]
Key is that we must launch with right-click and _Debug As > App Engine_ so that the execution proceeds via our `LaunchHelper` class.

It turns out that the devappserver hits a NPE very early on in startup (this seems to be a new behaviour change as I've haven't hit this NPE before).

```
Thread [main] (Suspended (exception NullPointerException))	
	owns: WeakCache$Factory  (id=62)	
	owns: Class<T> (com.google.appengine.api.appidentity.IAppIdentityServiceFactoryProvider) (id=63)	
	owns: Object  (id=64)	
	owns: Object  (id=65)	
	owns: Object  (id=66)	
	owns: Object  (id=67)	
	owns: IsolatedAppClassLoader  (id=68)	
	Transformer.isRuntimeCode(ProtectionDomain) line: 182	
	Transformer.transform(ClassLoader, String, Class<?>, ProtectionDomain, byte[]) line: 102	
	TransformerManager.transform(ClassLoader, String, Class, ProtectionDomain, byte[]) line: 188	
	InstrumentationImpl.transform(ClassLoader, String, Class, ProtectionDomain, byte[], boolean) line: 424	
	Proxy.defineClass0(ClassLoader, String, byte[], int, int) line: not available [native method]	
	Proxy.access$300(ClassLoader, String, byte[], int, int) line: 225	
	Proxy$ProxyClassFactory.apply(ClassLoader, Class<?>[]) line: 675	
	Proxy$ProxyClassFactory.apply(Object, Object) line: 592	
	WeakCache$Factory.get() line: 244	
	WeakCache<K,P,V>.get(K, P) line: 141	
	Proxy.getProxyClass0(ClassLoader, Class<?>...) line: 455	
	Proxy.newProxyInstance(ClassLoader, Class<?>[], InvocationHandler) line: 738	
	AnnotationParser.annotationForMap(Class<Annotation>, Map<String,Object>) line: 300	
	AnnotationParser.parseAnnotation2(ByteBuffer, ConstantPool, Class<?>, boolean, Class<Annotation>[]) line: 290	
	AnnotationParser.parseAnnotations2(byte[], ConstantPool, Class<?>, Class<Annotation>[]) line: 117	
	AnnotationParser.parseAnnotations(byte[], ConstantPool, Class<?>) line: 70	
	Class<T>.initAnnotationsIfNecessary() line: 3271	
	Class<T>.getAnnotation(Class<A>) line: 3219	
	IAppIdentityServiceFactoryProvider(FactoryProvider<I>).getPrecedence() line: 29	
	IAppIdentityServiceFactoryProvider(FactoryProvider<I>).hashCode() line: 46	
	IAppIdentityServiceFactoryProvider(Object).toString() line: 237	
	String.valueOf(Object) line: 2847	
	ServiceFactoryFactory$RuntimeRegistry.<clinit>() line: 115	
	ServiceFactoryFactory.getFactory(Class<T>) line: 67	
	DatastoreServiceFactory.getFactory() line: 72	
	DatastoreServiceFactory.getDatastoreService(DatastoreServiceConfig) line: 18	
	DatastoreServiceFactory.getDatastoreService() line: 38	
	TransactionCleanupFilter.getDatastoreService() line: 86	
	TransactionCleanupFilter.init(FilterConfig) line: 36	
	FilterHolder.doStart() line: 97	
	FilterHolder(AbstractLifeCycle).start() line: 50	
	ServletHandler.initialize() line: 662	
	DevAppEngineWebAppContext(Context).startContext() line: 140	
	DevAppEngineWebAppContext(WebAppContext).startContext() line: 1250	
	DevAppEngineWebAppContext(ContextHandler).doStart() line: 517	
	DevAppEngineWebAppContext(WebAppContext).doStart() line: 467	
	DevAppEngineWebAppContext(AbstractLifeCycle).start() line: 50	
	JettyContainerService$ApiProxyHandler(HandlerWrapper).doStart() line: 130	
	JettyContainerService$ApiProxyHandler(AbstractLifeCycle).start() line: 50	
	Server(HandlerWrapper).doStart() line: 130	
	Server.doStart() line: 224	
	Server(AbstractLifeCycle).start() line: 50	
	JettyContainerService.startContainer() line: 268	
	JettyContainerService(AbstractContainerService).startup() line: 284	
	AutomaticInstanceHolder.startUp() line: 26	
	AutomaticModule(AbstractModule<I>).startup() line: 87	
	Modules.startup() line: 105	
	DevAppServerImpl.doStart() line: 274	
	DevAppServerImpl.access$000(DevAppServerImpl) line: 47	
	DevAppServerImpl$1.run() line: 219	
```							

But when launching with a NPE Exception Breakpoint, this NPE throws execution back to the debugger.  But our `LaunchHelper`, which is running on the UI thread, is stuck waiting for the server to start.  Stack trace when hanging:
```
"Worker-397: Starting App Engine Standard at localhost (3)" #17529 prio=5 os_prio=31 tid=0x00007fe6bb7ad000 nid=0x19a8b in Object.wait() [0x0000700009e0b000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at org.eclipse.wst.server.core.internal.Server.startImpl(Server.java:3525)
	- locked <0x00000007b5d67188> (a [Z)
	at org.eclipse.wst.server.core.internal.Server$StartJob.run(Server.java:377)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:60)

"main" #1 prio=6 os_prio=31 tid=0x00007fe6bb011000 nid=0x307 in Object.wait() [0x00007ffeee3ad000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.eclipse.core.internal.jobs.Semaphore.acquire(Semaphore.java:39)
	- locked <0x00000007b5d919b0> (a org.eclipse.core.internal.jobs.Semaphore)
	at org.eclipse.core.internal.jobs.JobManager.join(JobManager.java:908)
	at org.eclipse.core.internal.jobs.InternalJob.join(InternalJob.java:341)
	at org.eclipse.core.runtime.jobs.Job.join(Job.java:537)
	at org.eclipse.wst.server.core.internal.Server.start(Server.java:2209)
	at org.eclipse.wst.server.core.internal.Server.start(Server.java:2126)
	at org.eclipse.wst.server.core.internal.Server.start(Server.java:2105)
	at com.google.cloud.tools.eclipse.appengine.localserver.launching.LaunchHelper.launch(LaunchHelper.java:142)
	at com.google.cloud.tools.eclipse.appengine.localserver.launching.LaunchHelper.launch(LaunchHelper.java:79)
	at com.google.cloud.tools.eclipse.appengine.localserver.launching.LocalAppEngineStandardLaunchShortcut.launch(LocalAppEngineStandardLaunchShortcut.java:44)
	at org.eclipse.debug.internal.ui.launchConfigurations.LaunchShortcutExtension.launch(LaunchShortcutExtension.java:431)
	at org.eclipse.debug.internal.ui.actions.LaunchShortcutAction.runInternal(LaunchShortcutAction.java:85)
	at org.eclipse.debug.internal.ui.actions.LaunchShortcutAction.runWithEvent(LaunchShortcutAction.java:135)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:565)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:397)
	at org.eclipse.jface.action.ActionContributionItem$$Lambda$207/562220412.handleEvent(Unknown Source)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:86)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4270)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1509)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1532)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1517)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1321)
	at org.eclipse.jface.action.ActionContributionItem.lambda$8(ActionContributionItem.java:1257)
	at org.eclipse.jface.action.ActionContributionItem$$Lambda$527/1971827603.handleEvent(Unknown Source)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:86)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4270)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1509)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1532)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1517)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1321)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4094)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3711)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1039)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:681)
	at org.eclipse.ui.internal.Workbench$$Lambda$24/1891834647.run(Unknown Source)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:595)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:655)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:591)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1501)
```

**Comment by aleguerrini** [comment]
If you can be useful I found the same problem on Eclipse Neon with the same configurations.

**Comment by briandealwis** [comment]
This hang is a bit unexpected as `IServer#start()` is documented as

> Asynchronously starts this server in the given launch mode.

Though the implementation `Server#start()` is more honest:
> Run the start logic with the synchronization flag taken from the server type

Our [`serverType`](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/blob/master/plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml#L48) is marked as `synchronousStart` as we do want any publishing to complete before starting the server: we only run with auto-start mode (restart a module when a change is published) when in RUN mode, as we don't want to restart modules in DEBUG when a change is made, since it kills the debugging thread.

**Comment by briandealwis** [comment]
@aleguerrini what other breakpoints do you have set?

**Comment by aleguerrini** [comment]
Only on the line:

`System.out.println("Print something.");`

but the problem is the same if add both a breakpoint on the line above:

`super.init();`

**Comment by aleguerrini** [comment]
I put loadOnStartup=1 to force the INIT servlet function call at the server startup.

**Comment by briandealwis** [comment]
So a workaround for now is to remove loadOnStartup.  That will avoid the breakpoint being triggered during the server launch.

**Comment by aleguerrini** [comment]
Yes. This is an example of code ... I discovered this problem because I have to debug on the initialization code of another Google App Engine project :-(

**Comment by aleguerrini** [comment]
In the meantime I discovered another bug ... try replacing the code of the INIT function with this code:
```
    @Override
    public void init() throws ServletException {

        super.init();

        if (true) {
              throw new RuntimeException();
        }

        System.out.println("Print something.");

    }

```

in this way I force an exception to the start. Try to click on **Run As -> App Engine** and try to **stop the server**. The server remains hung in "**STOPPING**" state.

**Comment by briandealwis** [comment]
That sounds like #1964.

A better workaround is to use the Debug drop-down in the toolbar, or the _Run > Debug History_ to launch the last configuration, or to use the _Servers_ view to launch the server.

The fix here is for `LaunchHelper` to call `IServer#start()` in a job.

**Comment by aleguerrini** [comment]
I updated the plugin to version 1.5.1 and finally I saw that the problem of slow startup, both in DEBUG and in RUN has been solved.

In any case I found that the following problem persists:

> In the meantime I discovered another bug ... try replacing the code of the INIT function with this code:
> ```
>     @Override
>     public void init() throws ServletException {
> 
>         super.init();
> 
>         if (true) {
>               throw new RuntimeException();
>         }
> 
>         System.out.println("Print something.");
> 
>     }
> 
> ```
> 
> in this way I force an exception to the start. Try to click on **Run As -> App Engine** and try to **stop the server**. The server remains hung in "**STOPPING**" state.
> 
> This means that if there is an exception in the startup code the server will not shut down anymore.

**Comment by chanseokoh** [comment]
@aleguerrini as @briandealwis said earlier, your last comment is another issue open for a while: #1964. The root cause is that the admin app at `/_ah/admin` is not working due to an error when loading your app.
